examples/grpc/src/main.rs

```diff
@@ -0,0 +1,162 @@
+use hyper::server::accept::from_stream;
+use hyper::service::make_service_fn;
+use hyper::{Client, Server, Uri};
+use std::convert::Infallible;
+use std::net::{IpAddr, Ipv4Addr};
+use tonic::Status;
+use tonic::{Request, Response};
+use tracing::{info_span, Instrument, Span};
+use turmoil::{net, Builder};
+
+mod proto {
+    tonic::include_proto!("helloworld");
+}
+
+use proto::greeter_server::{Greeter, GreeterServer};
+use proto::{HelloReply, HelloRequest};
+
+use crate::proto::greeter_client::GreeterClient;
+
+fn main() {
+    if std::env::var("RUST_LOG").is_err() {
+        std::env::set_var("RUST_LOG", "info");
+    }
+
+    tracing_subscriber::fmt::init();
+
+    let addr = (IpAddr::from(Ipv4Addr::UNSPECIFIED), 9999);
+
+    let mut sim = Builder::new().build();
+
+    let greeter = GreeterServer::new(MyGreeter {});
+
+    sim.host("server", move || {
+        let greeter = greeter.clone();
+        async move {
+            let listener = net::TcpListener::bind(addr).await.unwrap();
+
+            let accept = from_stream(async_stream::stream! {
+                loop {
+                    yield listener.accept().await.map(|(s, _)| s);
+                }
+            });
+
+            Server::builder(accept)
+                .serve(make_service_fn(move |_| {
+                    let greeter = greeter.clone();
+                    async move { Ok::<_, Infallible>(greeter) }.instrument(Span::current())
+                }))
+                .await
+                .unwrap();
+
+            Ok(())
+        }
+        .instrument(info_span!("server"))
+    });
+
+    sim.client(
+        "client",
+        async move {
+            let svc = Client::builder()
+                .http2_only(true)
+                .build(connector::connector());
+
+            let uri = Uri::from_static("http://server:9999");
+            let mut greeter_client = GreeterClient::with_origin(svc, uri);
+
+            let request = Request::new(HelloRequest { name: "foo".into() });
+
+            let res = greeter_client.say_hello(request).await?;
+
+            tracing::info!("Got response: {:?}", res);
+
+            Ok(())
+        }
+        .instrument(info_span!("client")),
+    );
+
+    sim.run().unwrap();
+}
+
+#[derive(Default)]
+pub struct MyGreeter {}
+
+#[tonic::async_trait]
+impl Greeter for MyGreeter {
+    async fn say_hello(
+        &self,
+        request: Request<HelloRequest>,
+    ) -> Result<Response<HelloReply>, Status> {
+        let reply = HelloReply {
+            message: format!("Hello {}!", request.into_inner().name),
+        };
+        Ok(Response::new(reply))
+    }
+}
+
+mod connector {
```

Comment:
What do you think about a hyper feature for this in turmoil? Premature maybe.

Reply:
I don't think the connection trait (the main reason we have to newtype) will be in hyper 1.0 so I am not sure if that is the right approach. Maybe adding a `turmoil-hyper` library could work.